src/runtime_src/core/common/drv/kds_cu_ctrl.c

```diff
@@ -0,0 +1,294 @@
+	if (unlikely(num_cus > kcuc->num_cus))
+		goto error;
+
+	/* If the configure command is sent by xclLoadXclbin(), the command
+	 * content should be the same and it is okay to let it go through.
+	 *
+	 * But it still has chance that user would manually construct a config
+	 * command, which could be wrong.
+	 *
+	 * So, do not allow reconfigure. This is still not totally safe, since
+	 * configure command and load xclbin are not atomic.
+	 *
+	 * The configured flag would be reset once the last one client finished.
+	 */
+	if (kcuc->configured) {
+		kds_info(client, "CU controller already configured\n");
+		goto done;
+	}
+
+	/* Now we need to make CU index right */
+	for (i = 0; i < num_cus; i++) {
+		j = get_cu_by_addr(kcuc, cus_addr[i]);
+		if (j == kcuc->num_cus)
+			goto error;
+
+		/* Ordering CU index */
+		if (j != i) {
+			tmp = kcuc->xcus[i];
+			kcuc->xcus[i] = kcuc->xcus[j];
+			kcuc->xcus[j] = tmp;
+		}
+		kcuc->xcus[i]->info.cu_idx = i;
+	}
+	kcuc->configured = 1;
+
+done:
+	mutex_unlock(&kcuc->lock);
+	xcmd->cb.notify_host(xcmd, KDS_COMPLETED);
+	xcmd->cb.free(xcmd);
+	return 0;
+
+error:
+	mutex_unlock(&kcuc->lock);
+	xcmd->cb.notify_host(xcmd, KDS_ERROR);
+	xcmd->cb.free(xcmd);
+	return -EINVAL;
+}
+
+int acquire_cu_inst_idx(struct kds_cu_ctrl *kcuc, struct kds_command *xcmd)
+{
+	struct kds_client *client = xcmd->client;
+	struct client_cu_priv *cu_priv;
+	int cu_idx;
+
+	/* Select CU */
+	cu_idx = cu_mask_to_cu_idx(xcmd);
+
+	/* Check if selected CU is in the context */
+	cu_priv = client->ctrl_priv[KDS_CU];
+	if (!test_bit(cu_idx, cu_priv->cu_bitmap)) {
+		xcmd->cb.notify_host(xcmd, KDS_ERROR);
+		xcmd->cb.free(xcmd);
+		return -EINVAL;
+	}
+
+	return kcuc->xcus[cu_idx]->info.inst_idx;
+}
+
+int control_ctx(struct kds_cu_ctrl *kcuc, struct kds_client *client,
+		struct kds_ctx_info *info)
+{
+	struct client_cu_priv *cu_priv;
+	u32 op;
+
+	/* TODO: Still has space to improve configured flag.
+	 * Since not all of the clients would need to use CU controller.
+	 *
+	 * But right now, the scope of a configuration is unclear.
+	 * Maybe the configuration could be per client?
+	 * Maybe config command would be removed?
+	 *
+	 * Anyway, for now, allow reconfigure when the last client exit.
+	 */
+	op = info->flags & CU_CTX_OP_MASK;
+	switch (op) {
+	case CU_CTX_OP_INIT:
+		cu_priv = vzalloc(sizeof(*cu_priv));
+		if (!cu_priv)
+			return -ENOMEM;
+		client->ctrl_priv[KDS_CU] = cu_priv;
+		mutex_lock(&kcuc->lock);
+		++kcuc->num_clients;
+		mutex_unlock(&kcuc->lock);
+		break;
+	case CU_CTX_OP_FINI:
+		vfree(client->ctrl_priv[KDS_CU]);
+		client->ctrl_priv[KDS_CU] = NULL;
+		mutex_lock(&kcuc->lock);
+		--kcuc->num_clients;
+		if (!kcuc->num_clients)
+			kcuc->configured = 0;
+		mutex_unlock(&kcuc->lock);
+		break;
+	case CU_CTX_OP_ADD:
+		return add_ctx(kcuc, client, info);
+	case CU_CTX_OP_DEL:
+		return del_ctx(kcuc, client, info);
+	}
+
+	return 0;
+}
+
+int add_cu(struct kds_cu_ctrl *kcuc, struct xrt_cu *xcu)
```

Comment:
I would suggest a clearer interface for the functions like this (remove_cu as well).
negative value means unsuccessful
0 means successful
if we need to return the index where the cu is added (or removed), we can add an extra pointer argument.

Reply:
Fixed